R/unique-keys.R

```diff
@@ -0,0 +1,250 @@
+
+dm_get_all_uks_impl <- function(dm, table = NULL) {
+  dm %>%
+    dm_get_def() %>%
+    dm_get_all_uks_def_impl(table)
+}
+
+dm_get_all_uks_def_impl <- function(def, table = NULL) {
+  # Optimized for speed
+
+  def_sub <- def[c("table", "uks")]
+
+  if (!is.null(table)) {
+    idx <- match(table, def_sub$table)
+    if (anyNA(idx)) {
+      abort(paste0("Table not in dm object: ", parent_table[which(is.na(idx))[[1]]]))
+    }
+    def_sub <- def_sub[match(table, def_sub$table), ]
+  }
+
+  out <-
+    def_sub %>%
+    unnest_df("uks", tibble(column = list())) %>%
+    set_names(c("table", "uk_col"))
+
+  out$uk_col <- new_keys(out$uk_col)
+  out
+}
+
+#' Remove a unique key
+#'
+#' @description
+#' `dm_rm_uk()` removes one or more unique keys from a table and leaves the [`dm`] object otherwise unaltered.
+#' An error is thrown if no unique key matches the selection criteria.
+#' If the selection criteria are ambiguous, a message with unambiguous replacement code is shown.
+#' Foreign keys are never removed.
+#'
+#' @inheritParams dm_rm_pk
+#' @param fail_fk
+#'   Boolean: if `TRUE` (default), will throw an error
+#'   if there are foreign keys addressing the unique key that is to be removed.
+#'
+#' @family primary key functions
+#'
+#' @return An updated `dm` without the indicated unique key(s).
+#'
+#' @export
+dm_rm_uk <- function(dm, table = NULL, columns = NULL, ..., fail_fk = TRUE) {
+  if (missing(fail_fk)) {
+    fail_fk <- NULL
+  }
+  dm_rm_uk_(dm, {{ table }}, {{ columns }}, ..., fail_fk = fail_fk)
+}
+
+dm_rm_uk_ <- function(dm, table, columns, ..., fail_fk = NULL) {
+  check_dots_empty()
+  check_not_zoomed(dm)
+
+  if (is.null(fail_fk)) {
+    fail_fk <- TRUE
+  }
+
+  table_name <- dm_tbl_name_null(dm, {{ table }})
+  columns <- enexpr(columns)
+
+  dm_rm_uk_impl(dm, table_name, columns, fail_fk)
+}
+
+dm_rm_uk_impl <- function(dm, table_name, columns, fail_fk) {
+  def <- dm_get_def(dm)
+
+  if (is.null(table_name)) {
+    i <- which(map_int(def$uks, vec_size) > 0)
+  } else {
+    i <- which(def$table == table_name)
+    if (nrow(def$uks[[i]]) == 0) {
+      i <- integer()
+    }
+  }
+
+  if (!quo_is_null(columns)) {
+    ii <- map2_lgl(def$data[i], def$uks[i], ~ tryCatch(
+      {
+        vars <- eval_select_indices(columns, colnames(.x))
+        identical(names(vars), .y$column[[1]])
+      },
+      error = function(e) {
+        FALSE
+      }
+    ))
+
+    i <- i[ii]
+  }
+
+  if (length(i) == 0 && dm_is_strict_keys(dm)) {
+    abort_uk_not_defined()
+  }
+
+  pwalk(list(def$fks[i], def$uks[i], def$table[i]), ~ {
+    is_match <- !is.na(vec_match(..1$ref_column, ..2$column))
+    if (fail_fk && any(is_match)) {
+      abort_first_rm_fks(..3, ..1$table[is_match])
+    }
+  })
+
+  # Talk about it
+  if (is.null(table_name)) {
+    message("Removing unique keys: %>%")
+    message("  ", glue_collapse(glue("dm_rm_uk({tick_if_needed(def$table[i])})"), " %>%\n  "))
+  }
+
+  # Execute
+  def$uks[i] <- list_of(new_uk())
```

Comment:
We need more work here -- this always removes all unique keys, we need to be more precise.

Reply:
excellent point, thanks, I'm on it.